pantab/__init__.py

```diff
@@ -14,3 +14,58 @@
     "frames_to_hyper",
     "test",
 ]
+
+# We link against HyperAPI in a fun way: In Python, we extract the function
+# pointers directly from the Python HyperAPI. We pass those function pointers
+# over to the C module which will then use those pointers to directly interact
+# with HyperAPI. Furthermore, we check the function signatures to guard
+# against API-breaking changes in HyperAPI.
+#
+# Directly using HyperAPI's C functions always was and still is discouraged and
+# unsupported by Tableu. In particular, Tableau will not be able to provide
+# official support for this hack.
+#
+# Because this is highly brittle, we try to make the error message as
+# actionable as possible and guide users in the right direction.
+
+api_incompatibility_msg = """
+pantab is incompatible with the version of Tableau Hyper API installed on your
+system. Please upgrade both `tableauhyperapi` and `pantab` to the latest version.
+If doing so does not fix this issue, please file an issue at
+https://github.com/innobi/pantab/issues mentioning the exact pantab and HyperAPI
+versions which triggered this error.
+"""
+
+
+def _check_signature(f, sig):
+    func_type = ffi.typeof(f)
+    assert func_type.kind == "function", f"expected a function, got {func_type.kind}"
+    assert func_type.cname == sig, f"expected '{sig}', got '{func_type.cname}'"
+    return f
+
+
+try:
+    from tableauhyperapi.impl.dll import lib, ffi
+
+    # fmt: off
```

Comment:
Rather than this can you just let black reformat this file?

Reply:
The things I really want to catch are:
1. if the import from `tableauhyperapi.impl.dll` fails, e.g. because we redesigned our internal package structure in HyperAPI
2. if any of the `lib.hyper_decode_date` properties does not exists (e.g. because we remove or rename a function)
3. if any of the `assert`s triggers

The easiest way to catch those 3 exception sources is to catch `AssertionError`, `ImportError` and `AttributeError`.

Is this sufficient, or would you prefer some other/more targeted error handling approach?

Reply:
You can see how this snippet looks if you let `black` reformat it here: https://github.com/vogelsgesang/pantab/commit/0f4a3595d7a4b292f060f4521515f0f2906cd9ff

`black` formats the different calls to `_get_hapi_function` in 3 ways:
1. on a single line
2. on 4 lines: function call in 1st line, each parameter on its own line, closing `)` on the last line 
3. on 3 lines: function call in 1st line, all parameters in 2nd line, closing `)` on the last line

I am completely fine with each of those 3 styles. But I dislike that `black` insists on mixing all 3 styles. Even if I manually add additional line breaks to consistently use style (2), i.e. break each call to 4 lines, `black` formats it back to a mixture of all 3 styles.

The semantics of this snippet is very "symmetric": Each call to `_get_hapi_function` does the exact same thing.
It's unfortunate that `black` insist on having 3 different styles for the individual calls and thereby hides this symmetry.
At least to me, the code gets less readable by having the exact same pattern being layouted in 3 different ways.

But in the end, I don't care too much...
Knowing how `black` wants to format this, do you still want me to apply `black` to those lines?

Reply:
Yea let's just stick with black's formatting

Reply:
done